cmd/kubelet/app/plugins_providers.go

```diff
@@ -69,9 +66,6 @@ type pluginInfo struct {
 
 func appendLegacyProviderVolumes(allPlugins []volume.VolumePlugin, featureGate featuregate.FeatureGate) ([]volume.VolumePlugin, error) {
 	pluginMigrationStatus := make(map[string]pluginInfo)
-	pluginMigrationStatus[plugins.AWSEBSInTreePluginName] = pluginInfo{pluginMigrationFeature: features.CSIMigrationAWS, pluginUnregisterFeature: features.InTreePluginAWSUnregister, pluginProbeFunction: awsebs.ProbeVolumePlugins}
-	pluginMigrationStatus[plugins.GCEPDInTreePluginName] = pluginInfo{pluginMigrationFeature: features.CSIMigrationGCE, pluginUnregisterFeature: features.InTreePluginGCEUnregister, pluginProbeFunction: gcepd.ProbeVolumePlugins}
-	pluginMigrationStatus[plugins.AzureDiskInTreePluginName] = pluginInfo{pluginMigrationFeature: features.CSIMigrationAzureDisk, pluginUnregisterFeature: features.InTreePluginAzureDiskUnregister, pluginProbeFunction: azuredd.ProbeVolumePlugins}
```

Comment:
this means the azure disk in-tree plugin does not register any more.
if there is an azure disk pv in k8s 1.23 cluster for example, then user upgrades to 1.27 directly, that in-tree pv would never work again unless user delete that pv and creates a csi driver pv.

I think even we remove the in-tree driver, those legacy in-tree drivers should still be registered, the driver requests would be forwarded to csi driver to handle.

cc @msau42 @jsafrane

Reply:
this issue also exists for gcepd and awsebs, is the above behavior wanted? IMO we should register those intree drivers as a plance holder driver, and csi driver would handle the in-tree requests finally. @jingxu97

Reply:
> if there is an azure disk pv in k8s 1.23 cluster for example, then user upgrades to 1.27 directly...

FWIW, skip-level control-plane upgrades are _explicitly_ unsupported. see https://kubernetes.io/releases/version-skew-policy/#kube-apiserver-1

>> Note: Project policies for [API deprecation](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) and [API change guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) require kube-apiserver to not skip minor versions when upgrading, even in single-instance clusters.

> ...that in-tree pv would never work again unless user delete that pv and creates a csi driver pv.

what process between 1.23 and 1.27 handles the in-tree → CSI migration?

Reply:
it would also break if upgrades 1.26 to 1.27 for in-tree pv, as long as it's on 1.27 with this PR, in-tree pv does not work any more since the legacy in-tree driver name is not registered any more.

Reply:
In a cluster with kube-apiserver and kube-controller-manager 1.27, there can be kubelets with ver. 1.27, 1.26 and 1.25 (see [version skew policy](https://kubernetes.io/releases/version-skew-policy/#kubelet)). CSI migration of AzureDisk, GCE PD and AWS EBS [are GA in 1.25](https://github.com/kubernetes/kubernetes/blob/3534f896e9ea6936b50d821954374374f41c58bd/pkg/features/kube_features.go#L917).

Therefore all nodes in the cluster + KCM + API server have CSI migration GA and in-tree volume plugins are not registered in any component.

AzureFile + vSphere CSI driver feature gates must not be removed, they were GA in 1.26. And they are not removed in this PR.

Reply:
> it would also break if upgrades from 1.26 to 1.27 for in-tree pv, as long as it's on 1.27 with this PR, in-tree pv does not work any more since the legacy in-tree driver name is not registered any more.

In-tree PVs are quietly translated to CSI and attached/mounted by the CSI driver. This translation is in-memory only, the PVs still look like in-tree in the API server.

Reply:
thanks, I'm not sure about this part of the code, so we should keep these changes here?

Reply:
I think the PR is OK, at least now.

Reply:
Thanks @jsafrane